File: plugins/pangle_custom_plugin/ios/Classes/PangleNativeHandler.swift
import Flutter
import PAGAdSDK
import UIKit

/// Pangle iOS 네이티브 구현
public class PangleNativeHandler: NSObject, FlutterPlugin {
    private static var channel: FlutterMethodChannel?
    private var rewardedAd: PAGRewardedAd?
    private var isSDKInitialized = false
    private var appID: String?

    /// Flutter 플러그인 등록
    public static func register(with registrar: FlutterPluginRegistrar) {
        print("PangleNativeHandler 등록 시작")
        channel = FlutterMethodChannel(
            name: "pangle_native_channel",
            binaryMessenger: registrar.messenger())
        let instance = PangleNativeHandler()
        registrar.addMethodCallDelegate(instance, channel: channel!)
        print("PangleNativeHandler 등록 완료")
    }

    /// Flutter 메서드 호출 처리
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("메서드 호출 수신: \(call.method), 인자: \(String(describing: call.arguments))")

        switch call.method {
        case "initPangle":
            if let args = call.arguments as? [String: Any],
                let appId = args["appId"] as? String
            {
                print("Pangle 초기화 시작 - appId: \(appId)")
                self.appID = appId
                self.initPangle(appId: appId) { success, errorMessage in
                    if success {
                        print("Pangle 초기화 성공 - 결과 리턴")
                        result(true)
                    } else {
                        print("Pangle 초기화 실패 - 에러: \(errorMessage)")
                        result(
                            FlutterError(code: "InitFailed", message: errorMessage, details: nil))
                    }
                }
            } else {
                print("Pangle 초기화 실패 - 인자 없음")
                result(FlutterError(code: "InvalidParams", message: "App ID is null", details: nil))
            }
        case "loadRewardedAd":
            if let args = call.arguments as? [String: Any],
                let placementId = args["placementId"] as? String
            {
                if self.isSDKInitialized {
                    self.loadRewardedAd(placementId: placementId, result: result)
                } else if let appId = self.appID {
                    // SDK가 초기화되지 않았다면 다시 초기화 시도 후 광고 로드
                    print("SDK가 초기화되지 않았습니다. 재초기화 시도 중...")
                    self.initPangle(appId: appId) { success, errorMessage in
                        if success {
                            self.loadRewardedAd(placementId: placementId, result: result)
                        } else {
                            result(
                                FlutterError(
                                    code: "InitFailed",
                                    message: "Pangle SDK 초기화 실패: \(errorMessage)", details: nil))
                        }
                    }
                } else {
                    result(
                        FlutterError(
                            code: "NotInitialized", message: "Pangle SDK가 초기화되지 않았습니다", details: nil
                        ))
                }
            } else {
                result(
                    FlutterError(
                        code: "InvalidParams", message: "Placement ID is null", details: nil))
            }
        case "showRewardedAd":
            self.showRewardedAd(result: result)
        default:
            print("지원하지 않는 메서드: \(call.method)")
            result(FlutterMethodNotImplemented)
        }
    }
}

/// Pangle 광고 관련 기능 구현
extension PangleNativeHandler: PAGRewardedAdDelegate {
    func initPangle(appId: String, completion: @escaping (Bool, String) -> Void) {
        print("initPangle 메서드 호출됨 - appId: \(appId)")

        // 이미 초기화된 상태면 바로 성공 반환
        if isSDKInitialized && self.appID == appId {
            print("Pangle SDK가 이미 초기화되어 있습니다.")
            completion(true, "이미 초기화됨")
            return
        }

        print("PAGConfig 생성 및 설정...")
        let config = PAGConfig.share()
        config.appID = appId

        // 디버그 모드 활성화 및 추가 설정
        #if DEBUG
            print("디버그 모드 활성화")
        #endif

        // 메인 스레드에서 초기화 확실히 보장
        DispatchQueue.main.async {
            print("PAGSdk.start 호출 (메인 스레드)...")
            PAGSdk.start(with: config) { [weak self] success, error in
                guard let self = self else {
                    print("self가 nil이라 초기화 완료 처리 불가")
                    completion(false, "객체가 해제됨")
                    return
                }

                if success {
                    print("Pangle SDK 초기화 성공")
                    self.isSDKInitialized = true
                    self.appID = appId

                    // Pangle SDK 상태 확인
                    print("Pangle SDK 초기화 상태: \(self.isSDKInitialized)")
                    print("Pangle SDK appID: \(self.appID ?? "nil")")

                    completion(true, "성공")
                } else {
                    let errorMessage = error?.localizedDescription ?? "알 수 없는 오류"
                    print("Pangle SDK 초기화 실패: \(errorMessage)")
                    print("에러 상세: \(String(describing: error))")
                    self.isSDKInitialized = false
                    completion(false, errorMessage)
                }
            }
        }
    }

    func loadRewardedAd(placementId: String, result: @escaping FlutterResult) {
        print("리워드 광고 로드 시작 - placementId: \(placementId)")

        // SDK가 초기화되지 않았으면 초기화 상태를 확인
        if !isSDKInitialized {
            print("Pangle SDK가 초기화되지 않았습니다. 초기화 필요")
            result(
                FlutterError(
                    code: "NotInitialized", message: "Pangle SDK가 초기화되지 않았습니다", details: nil))
            return
        }

        // 이전 광고 객체 정리
        self.rewardedAd = nil

        let request = PAGRewardedRequest()

        // 슬롯 ID 로그 출력
        print("PAGRewardedAd.load 호출 준비 - placementId: \(placementId)")

        // 안전 장치: 광고 로드 시도 전 약간의 지연 추가
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            // 메인 스레드에서 로드 시작
            PAGRewardedAd.load(withSlotID: placementId, request: request) {
                [weak self] (rewardedAd: PAGRewardedAd?, error: Error?) in
                // 메인 스레드에서 콜백 처리 보장
                DispatchQueue.main.async {
                    guard let self = self else {
                        print("self가 nil이라 광고 로드 처리 불가")
                        result(FlutterError(code: "LoadFailed", message: "객체가 해제됨", details: nil))
                        return
                    }

                    if let error = error {
                        print("리워드 광고 로드 실패: \(error.localizedDescription)")
                        print("오류 상세: \(error)")
                        result(
                            FlutterError(
                                code: "LoadFailed", message: error.localizedDescription,
                                details: nil))
                        return
                    }

                    if let rewardedAd = rewardedAd {
                        print("리워드 광고 로드 성공: \(rewardedAd)")
                        self.rewardedAd = rewardedAd
                        self.rewardedAd?.delegate = self
                        // 로드 성공 여부 다시 한번 확인
                        if self.rewardedAd != nil {
                            print("광고 객체 저장 성공")
                            result(true)
                        } else {
                            print("광고 객체 저장 실패")
                            result(
                                FlutterError(
                                    code: "LoadFailed", message: "광고 객체 저장 실패", details: nil))
                        }
                    } else {
                        print("리워드 광고 로드 실패: 광고 객체가 null임")
                        result(
                            FlutterError(
                                code: "LoadFailed", message: "알 수 없는 오류로 광고 로드 실패", details: nil))
                    }
                }
            }
        }
    }

    func showRewardedAd(result: @escaping FlutterResult) {
        guard let rewardedAd = rewardedAd else {
            print("리워드 광고 표시 실패: 광고 객체가 nil입니다")
            result(FlutterError(code: "ShowFailed", message: "리워드 광고가 준비되지 않았습니다", details: nil))
            return
        }

        print("정확한 rootViewController 찾기 시도 중...")

        // rootViewController를 더 정확하게 찾기
        guard let keyWindow = UIApplication.shared.windows.first(where: { $0.isKeyWindow }) else {
            print("리워드 광고 표시 실패: keyWindow를 찾을 수 없습니다")
            result(FlutterError(code: "ShowFailed", message: "keyWindow를 찾을 수 없습니다", details: nil))
            return
        }

        guard let rootVC = keyWindow.rootViewController else {
            print("리워드 광고 표시 실패: rootViewController를 찾을 수 없습니다")
            result(
                FlutterError(
                    code: "ShowFailed", message: "rootViewController를 찾을 수 없습니다", details: nil))
            return
        }

        // 가장 상위의 표시 중인 뷰컨트롤러 찾기
        var topVC = rootVC
        while let presentedVC = topVC.presentedViewController {
            topVC = presentedVC
        }

        print("광고 표시에 사용할 뷰컨트롤러: \(topVC)")

        // 메인 스레드에서 확실하게 실행
        DispatchQueue.main.async {
            print("리워드 광고 표시 시작")

            // 기존 광고가 표시 중인지 확인 후 표시
            rewardedAd.delegate = self

            // 광고 표시 전에 delegate를 다시 한번 명시적으로 설정
            if let ad = self.rewardedAd {
                ad.delegate = self
                ad.present(fromRootViewController: topVC)
                print("present(fromRootViewController:) 메서드 호출됨")
                result(true)
            } else {
                print("리워드 광고 표시 실패: 마지막 순간에 광고 객체가 nil이 됨")
                result(FlutterError(code: "ShowFailed", message: "광고 객체가 nil이 되었습니다", details: nil))
            }
        }
    }

    // MARK: - PAGRewardedAdDelegate

    func rewardedAdDidShow(_ rewardedAd: PAGRewardedAd) {
        print("리워드 광고가 표시됨: \(rewardedAd)")
        PangleNativeHandler.channel?.invokeMethod("onAdShowed", arguments: nil)
    }

    func rewardedAdDidClick(_ rewardedAd: PAGRewardedAd) {
        print("리워드 광고가 클릭됨: \(rewardedAd)")
        PangleNativeHandler.channel?.invokeMethod("onAdClicked", arguments: nil)
    }

    func rewardedAdDidClose(_ rewardedAd: PAGRewardedAd) {
        print("리워드 광고가 닫힘: \(rewardedAd)")
        self.rewardedAd = nil
        PangleNativeHandler.channel?.invokeMethod("onAdClosed", arguments: nil)
    }

    func rewardedAd(_ rewardedAd: PAGRewardedAd, didRewardSuccess rewardInfo: PAGRewardModel) {
        print("사용자가 보상을 받음: \(rewardInfo)")
        let rewardData: [String: Any] = [
            "amount": rewardInfo.rewardAmount,
            "name": rewardInfo.rewardName,
        ]
        PangleNativeHandler.channel?.invokeMethod("onUserEarnedReward", arguments: rewardData)
    }
}

File: picnic_app/android/app/src/main/kotlin/io/iconcasting/picnic/app/pangle/PangleNativeHandler.kt
package pangle.custom

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import com.bytedance.sdk.openadsdk.api.init.PAGConfig
import com.bytedance.sdk.openadsdk.api.init.PAGSdk
import com.bytedance.sdk.openadsdk.api.reward.PAGRewardItem
import com.bytedance.sdk.openadsdk.api.reward.PAGRewardedAd
import com.bytedance.sdk.openadsdk.api.reward.PAGRewardedAdInteractionListener
import com.bytedance.sdk.openadsdk.api.reward.PAGRewardedAdLoadListener
import com.bytedance.sdk.openadsdk.api.reward.PAGRewardedRequest
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** Pangle Android 네이티브 구현 */
class PangleNativeHandler : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private lateinit var applicationContext: Context
    private var activity: Activity? = null
    private var rewardedAd: PAGRewardedAd? = null
    private var isSDKInitialized = false
    private var appID: String? = null
    private var sandboxPlacementId: String? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "pangle_native_channel")
        channel.setMethodCallHandler(this)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initPangle" -> {
                val appId = call.argument<String>("appId")
                val userId = call.argument<String>("userId")
                val environment = call.argument<String>("environment") ?: "prod"
                val productionAppId = call.argument<String>("productionAppId")
                val requestedSandboxPlacement = call.argument<String>("sandboxPlacementId")
                val productionPlacementId = call.argument<String>("productionPlacementId")
                val sandboxConfigValid = environment != "sandbox" ||
                    (appId != null &&
                        appId.isNotEmpty() &&
                        productionAppId != null &&
                        appId != productionAppId &&
                        !requestedSandboxPlacement.isNullOrEmpty() &&
                        requestedSandboxPlacement != productionPlacementId)
                if (appId != null &&
                    environment in setOf("prod", "sandbox") &&
                    sandboxConfigValid
                ) {
                    appID = appId
                    sandboxPlacementId =
                        if (environment == "sandbox") requestedSandboxPlacement else null
                    initPangle(appId, userId, result)
                } else {
                    result.error("InvalidParams", "Valid SDK configuration is required", null)
                }
            }
            "loadRewardedAd" -> {
                val placementId = call.argument<String>("placementId")
                    ?: return result.error("InvalidParams", "placementId is required", null)
                if (sandboxPlacementId != null && placementId != sandboxPlacementId) {
                    return result.error(
                        "InvalidSandboxPlacement",
                        "Sandbox placement rejected",
                        null,
                    )
                }
                val mediaExtra = call.argument<String>("mediaExtra")
                    ?: return result.error("InvalidParams", "mediaExtra is required", null)
                try {
                    val validated = PangleMediaExtra.requireV2(mediaExtra)
                    val configuredAppId = appID
                    if (!isSDKInitialized && configuredAppId == null) {
                        result.error("NotInitialized", "Pangle SDK가 초기화되지 않았습니다", null)
                        return
                    }
                    PangleLoadCoordinator.run(
                        initialized = isSDKInitialized,
                        initialize = { completion ->
                            initPangle(
                                configuredAppId!!,
                                null,
                                completion = completion,
                            )
                        },
                        placementId = placementId,
                        mediaExtra = validated,
                        load = { id, extra -> loadRewardedAd(id, extra, result) },
                        fail = { error -> result.error("InitFailed", error.message, null) },
                    )
                } catch (_: IllegalArgumentException) {
                    result.error("InvalidMediaExtra", "Signed v2 mediaExtra is required", null)
                }
            }
            "showRewardedAd" -> {
                showRewardedAd(result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun initPangle(
        appId: String,
        userId: String?,
        flutterResult: Result? = null,
        completion: ((kotlin.Result<Unit>) -> Unit)? = null,
    ) {
        println("Flutter에서 Pangle SDK 초기화 시작")

        if (isSDKInitialized && appID == appId) {
            println("Pangle SDK가 이미 초기화되어 있습니다.")
            flutterResult?.success(true)
            completion?.invoke(kotlin.Result.success(Unit))
            return
        }

        val pagConfig = PAGConfig.Builder()
            .appId(appId)
            .debugLog(true)
            .build()

        PAGSdk.init(applicationContext, pagConfig, object : PAGSdk.PAGInitCallback {
            override fun success() {
                println("Pangle SDK 초기화 성공")
                isSDKInitialized = true
                appID = appId
                flutterResult?.success(true)
                completion?.invoke(kotlin.Result.success(Unit))
            }

            override fun fail(code: Int, msg: String) {
                println("Pangle SDK 초기화 실패: $msg (코드: $code)")
                isSDKInitialized = false
                flutterResult?.error("InitFailed", msg, null)
                completion?.invoke(kotlin.Result.failure(IllegalStateException(msg)))
            }
        })
    }

    private fun loadRewardedAd(placementId: String, mediaExtra: String, result: Result) {
        println("리워드 광고 로드 시작 - placementId: $placementId")

        rewardedAd = null
        val request = PAGRewardedRequest()
        val extraInfo = hashMapOf<String, Any>()
        extraInfo["media_extra"] = mediaExtra
        request.extraInfo = extraInfo

        Handler(Looper.getMainLooper()).postDelayed({
            PAGRewardedAd.loadAd(placementId, request, object : PAGRewardedAdLoadListener {
                override fun onError(code: Int, msg: String) {
                    println("리워드 광고 로드 실패: $msg (코드: $code)")
                    result.error("LoadFailed", msg, null)
                }

                override fun onAdLoaded(ad: PAGRewardedAd) {
                    println("리워드 광고 로드 성공")
                    rewardedAd = ad
                    result.success(true)
                }
            })
        }, 500)
    }

    private fun showRewardedAd(result: Result) {
        if (rewardedAd != null) {
            println("리워드 광고 표시 시작")

            val currentActivity = activity
            if (currentActivity == null) {
                result.error("ShowFailed", "Activity가 없습니다", null)
                return
            }

            rewardedAd?.setAdInteractionListener(object : PAGRewardedAdInteractionListener {
                override fun onAdShowed() {
                    println("리워드 광고가 표시됨")
                    channel.invokeMethod(PangleEventNames.AD_SHOWN, null)
                }

                override fun onAdClicked() {
                    println("리워드 광고가 클릭됨")
                    channel.invokeMethod(PangleEventNames.AD_CLICKED, null)
                }

                override fun onAdDismissed() {
                    println("리워드 광고가 닫힘")
                    rewardedAd = null
                    channel.invokeMethod(PangleEventNames.AD_DISMISSED, null)
                }

                override fun onUserEarnedReward(item: PAGRewardItem) {
                    println("사용자가 보상을 받음: ${item.rewardAmount} ${item.rewardName}")
                    val rewardData = mapOf(
                        "amount" to item.rewardAmount,
                        "name" to item.rewardName
                    )
                    channel.invokeMethod(PangleEventNames.REWARD_EARNED, rewardData)
                }

                override fun onUserEarnedRewardFail(code: Int, msg: String) {
                    println("사용자 보상 획득 실패: $msg (코드: $code)")
                    val errorData = mapOf(
                        "code" to code,
                        "errorMessage" to msg
                    )
                    channel.invokeMethod(PangleEventNames.REWARD_FAILED, errorData)
                }
            })

            try {
                rewardedAd?.show(currentActivity)
                result.success(true)
            } catch (e: Exception) {
                println("광고 표시 중 예외 발생: ${e.message}")
                result.error("ShowFailed", "광고 표시 중 오류 발생: ${e.message}", null)
            }
        } else {
            result.error("ShowFailed", "리워드 광고가 준비되지 않았습니다", null)
        }
    }
}
